rust/noosphere-sphere/src/authority/read.rs

```diff
@@ -0,0 +1,162 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use noosphere_core::{
+    authority::{Author, Authorization},
+    data::{Did, Link, Mnemonic},
+};
+use noosphere_storage::Storage;
+
+use tokio_stream::StreamExt;
+
+use crate::{HasSphereContext, SphereContextKey};
+use async_trait::async_trait;
+
+/// Anything that can read the authority section from a sphere should implement
+/// [SphereAuthorityRead]. A blanket implementation is provided for anything
+/// that implements [HasSphereContext].
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait SphereAuthorityRead<S>
+where
+    S: Storage + 'static,
+    Self: Sized,
+{
+    /// For a given [Authorization], checks that the authorization and all of its
+    /// ancester proofs are valid and have not been revoked
+    async fn verify_authorization(&self, authorization: &Authorization) -> Result<()>;
+
+    /// Look up an authorization by a [Did].
+    async fn get_authorization(&self, did: &Did) -> Result<Option<Authorization>>;
+
+    /// Derive a root sphere key from a mnemonic and return a version of this
+    /// [SphereAuthorityRead] whose inner [SphereContext]'s [Author] is using
+    /// that root sphere key.
+    async fn escalate_authority(&self, mnemonic: &Mnemonic) -> Result<Self>;
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<C, S> SphereAuthorityRead<S> for C
+where
+    C: HasSphereContext<S>,
+    S: Storage + 'static,
+{
+    async fn verify_authorization(&self, authorization: &Authorization) -> Result<()> {
+        self.to_sphere()
+            .await?
+            .verify_authorization(authorization)
+            .await
+    }
+
+    async fn get_authorization(&self, did: &Did) -> Result<Option<Authorization>> {
+        let sphere = self.to_sphere().await?;
+        let authority = sphere.get_authority().await?;
+        let delegations = authority.get_delegations().await?;
+        let delegations_stream = delegations.into_stream().await?;
+
+        tokio::pin!(delegations_stream);
+
+        while let Some((Link { cid, .. }, delegation)) = delegations_stream.try_next().await? {
+            let ucan = delegation.resolve_ucan(sphere.store()).await?;
+            let authorized_did = ucan.audience();
+
+            if authorized_did == did {
```

Comment:
We should shore up delegation so that we don't delegate to another key more than once. In practice I don't think we have a use case for multiple levels of access via different delegations.